### PR TITLE
Feature validation

### PR DESCRIPTION
**Pull request by jay3p**
- Adding validation
  - Dataset will be split into a training and a validation dataset
  - size of both sets configurable in config.py
- Adding early stopping
  - Training stops when validation loss doesn't improve
- Adding loss weights
  - Useful to bias the training process towards specific features 
- Adding gradient clipping
  - Saves the training process from being corrupted by exploding gradients (e.g. because of numerical instability)

### CODE REVIEW

**Comment by julienpoeschl** [comment]
Ist early stopping auch abschaltbar?

**Comment by jay3p** [comment]
Early stopping kann man jetzt auch in der config deaktivieren (wobei das aus meiner Sicht in den meisten Fällen nicht so sinnvoll ist) und anpassen
Feature Temp kann ich später noch adden.

**Comment by julienpoeschl** [comment]
Alles klar. Mein Gedanke war, dass man eventuel für die Präsentation zeigen möchte, was passiert wenn man nicht abbricht. Sonst macht das natürlich Sinn. Und tuple für temp habe ich gefragt, da man alle features ihre temp regulieren können sollte, und je nach späteren Änderungen vllt Bar temp doch wieder verändern möchte

**Comment by julienpoeschl** [comment]
Ich lese mir nachher den code durch

File: src/groove_panda/config.py [hunk 1]
```diff
@@ -173,7 +183,7 @@ class Parser(Enum):
             "pitch": 128,
             "duration": 64,
             "velocity": 64,
-            "position": 64,
+            "position": 32,
```
Comment (julienpoeschl) [inline src/groove_panda/config.py:186]: Ist diese Änderung Teil eines Presets, dass wir vielleicht erhalten wollen? Also ob wir hier das Preset für alle verändern wollen

File: src/groove_panda/generation/generate.py [hunk 2]
```diff
@@ -56,8 +56,8 @@ def _predictions_to_sixtuple(self, predictions: ModelPredictions) -> Sixtuple:
         """
 
         # Apply temperature sampling to each feature prediction
-        bar_idx = self._apply_temperature_sampling(predictions.bar[0], self.temperature)
-        position_idx = self._apply_temperature_sampling(predictions.position[0], self.temperature)
+        bar_idx = self._apply_temperature_sampling(predictions.bar[0], 0)
```
Comment (julienpoeschl) [inline src/groove_panda/generation/generate.py:59]: Hier fände ich persönlich die Option, die temp für jedes feature selbst zu wählen, über config eintrag, ganz angenehm. Auch wenn man es nicht verändert, gibt dass config dann ganz nice übersicht über alle dinge, die wir irgendwie festlegen, um das model zu verbessern/verändern

File: src/groove_panda/models/train.py [hunk 3]
```diff
@@ -86,49 +94,89 @@ def train_model_eager(model: BaseModel, train_generator: FlexibleSequenceGenerat
...
             for idx, feature in enumerate(FEATURE_NAMES)
         }
 
-        dataset = tf.data.Dataset.from_tensor_slices((x_dict, y_dict))
+        logger.info("Giving dataset to TensorFlow...")
+        train_dataset = tf.data.Dataset.from_tensor_slices((train_x_dict, train_y_dict))
+        val_dataset = tf.data.Dataset.from_tensor_slices((val_x_dict, val_y_dict))
 
-        logger.info("Start shuffling...")
         # Providing input for the model is now handled by Tensorflow since it's maximally optimized
```
Comment (julienpoeschl) [inline src/groove_panda/models/train.py:137]: Multiline comments sind meistens für so große erklärungen etwas praktischer. Dann kann man den text auch leichter in zukunft modifizieren

File: src/groove_panda/models/utils.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
+def get_loss_weights():
```
Comment (julienpoeschl) [inline src/groove_panda/models/utils.py:1]: Return type wäre angenehm hier

Comment (julienpoeschl) [reply src/groove_panda/models/utils.py:1]: Falls du es noch nicht gemacht hast: VSCode -> Settings -> Suchleiste: type checking -> unter der feature matrix im dropdown auf basic stellen

File: src/groove_panda/config.py (hunk 1 above)
Comment (jay3p) [reply src/groove_panda/config.py:186]: Damn, das ist da reingerutscht

File: src/groove_panda/models/train.py (hunk 3 above)
Comment (jay3p) [reply src/groove_panda/models/train.py:137]: Ändere ich ggf.
Habe nur gesehen, dass das vom Interpreter anders behandelt wird und deshalb evtl nicht so optimal ist. But i'll look into it.

Comment (julienpoeschl) [reply src/groove_panda/models/train.py:137]: Interesting. Was macht der damit? Versucht der die wie methoden docs zu behandeln?

File: src/groove_panda/config.py (hunk 1 above)
Comment (julienpoeschl) [reply src/groove_panda/config.py:186]: Da mit Joao's Änderungen das eh egal ist, schließe ich diese konversation

File: src/groove_panda/generation/generate.py (hunk 2 above)
Comment (julienpoeschl) [reply src/groove_panda/generation/generate.py:59]: Das macht irgendjemand diese Woche

File: src/groove_panda/models/train.py (hunk 3 above)
Comment (jay3p) [reply src/groove_panda/models/train.py:137]: Das nicht, aber zur Laufzeit werden die wohl meist als Strings in den Programmspeicher geladen. Also sollte man die nur für Docstrings verwenden (habe ich leider auch erst vor kurzem gelernt). Ist die Frage, wie wir damit umgehen

Comment (julienpoeschl) [reply src/groove_panda/models/train.py:144]: Ist glaube ich relativ egal für uns, oder? Aber für nächstes python projekt merken
